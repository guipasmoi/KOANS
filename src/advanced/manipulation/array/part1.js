export function example1(console) {
  const names = ["Ben", "Jafar", "Matt", "Priya", "Brian"];
  let counter = 0;

  for (counter = 0; counter < names.length; counter++) {
    console.log(names[counter]);
  }
}

export function example2(console) {
  const names = ["Ben", "Jafar", "Matt", "Priya", "Brian"];

  names.forEach(name => {
    console.log(name);
  });
}

export function exo1() {
  const newReleases = [
      {
        id: 70111470,
        title: "Die Hard",
        boxart: "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [4.0],
        bookmark: []
      },
      {
        id: 654356453,
        title: "Bad Boys",
        boxart: "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [5.0],
        bookmark: [{ id: 432534, time: 65876586 }]
      },
      {
        id: 65432445,
        title: "The Chamber",
        boxart: "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [4.0],
        bookmark: []
      },
      {
        id: 675465,
        title: "Fracture",
        boxart: "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [5.0],
        bookmark: [{ id: 432534, time: 65876586 }]
      }
    ],
    videoAndTitlePairs = [];

  // ------------ INSERT CODE HERE! -----------------------------------
  // Use forEach function to accumulate {id, title} pairs from each video.
  // Put the results into the videoAndTitlePairs array using the Array's
  // push() method. Example: videoAndTitlePairs.push(newItem);
  newReleases.forEach(video => {
    videoAndTitlePairs.push({ id: video.id, title: video.title });
  });
  // ------------ INSERT CODE HERE! -----------------------------------

  return videoAndTitlePairs;
}

export function exo2() {
  Array.prototype.map = function(projectionFunction) {
    const results = [];
    this.forEach(function(itemInArray) {
      // ------------ INSERT CODE HERE! ----------------------------
      // Apply the projectionFunction to each item in the array and add
      // each result to the results array.
      // Note: you can add items to an array with the push() method.
      results.push(projectionFunction(itemInArray));
      // ------------ INSERT CODE HERE! ----------------------------
    });

    return results;
  };
}

export function exo3() {
  const newReleases = [
      {
        id: 70111470,
        title: "Die Hard",
        boxart: "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [4.0],
        bookmark: []
      },
      {
        id: 654356453,
        title: "Bad Boys",
        boxart: "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [5.0],
        bookmark: [{ id: 432534, time: 65876586 }]
      },
      {
        id: 65432445,
        title: "The Chamber",
        boxart: "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [4.0],
        bookmark: []
      },
      {
        id: 675465,
        title: "Fracture",
        boxart: "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
        uri: "http://api.netflix.com/catalog/titles/movies/70111470",
        rating: [5.0],
        bookmark: [{ id: 432534, time: 65876586 }]
      }
    ],
    videoAndTitlePairs = [];

  // ------------ INSERT CODE HERE! -----------------------------------
  // Use forEach function to accumulate {id, title} pairs from each video.
  // Put the results into the videoAndTitlePairs array using the Array's
  // push() method. Example: videoAndTitlePairs.push(newItem);
  newReleases.forEach(video => {
    videoAndTitlePairs.push({ id: video.id, title: video.title });
  });
  // ------------ INSERT CODE HERE! -----------------------------------

  return videoAndTitlePairs;
}

export function exo4() {
  Array.prototype.map = function(projectionFunction) {
    const results = [];
    this.forEach(function(itemInArray) {
      // ------------ INSERT CODE HERE! ----------------------------
      // Apply the projectionFunction to each item in the array and add
      // each result to the results array.
      // Note: you can add items to an array with the push() method.
      results.push(projectionFunction(itemInArray));
      // ------------ INSERT CODE HERE! ----------------------------
    });

    return results;
  };
  // JSON.stringify([1,2,3].map(function(x) { return x + 1; })) === '[2,3,4]'
}

export function exo5() {
  const newReleases = [
    {
      id: 70111470,
      title: "Die Hard",
      boxart: "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: [4.0],
      bookmark: []
    },
    {
      id: 654356453,
      title: "Bad Boys",
      boxart: "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: [5.0],
      bookmark: [{ id: 432534, time: 65876586 }]
    },
    {
      id: 65432445,
      title: "The Chamber",
      boxart: "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: [4.0],
      bookmark: []
    },
    {
      id: 675465,
      title: "Fracture",
      boxart: "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: [5.0],
      bookmark: [{ id: 432534, time: 65876586 }]
    }
  ];

  // ------------ INSERT CODE HERE! -----------------------------------
  // Use map function to accumulate {id, title} pairs from each video.
  //return newReleases.map; // finish this expression!
  return newReleases.map(video => ({ id: video.id, title: video.title }));
  // ------------ INSERT CODE HERE! -----------------------------------
}

export function exo6() {
  const newReleases = [
    {
      id: 70111470,
      title: "Die Hard",
      boxart: "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 4.0,
      bookmark: []
    },
    {
      id: 654356453,
      title: "Bad Boys",
      boxart: "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 5.0,
      bookmark: [{ id: 432534, time: 65876586 }]
    },
    {
      id: 65432445,
      title: "The Chamber",
      boxart: "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 4.0,
      bookmark: []
    },
    {
      id: 675465,
      title: "Fracture",
      boxart: "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 5.0,
      bookmark: [{ id: 432534, time: 65876586 }]
    }
  ];
  const videos = [];

  // ------------ INSERT CODE HERE! -----------------------------------
  // Use forEach function to accumulate every video with a rating of 5.0
  newReleases.forEach(function(video) {
    if (video.rating === 5.0) {
      videos.push(video);
    }
  });
  // ------------ INSERT CODE HERE! -----------------------------------

  return videos;
}

export function exo7() {
  Array.prototype.filter = function(predicateFunction) {
    const results = [];
    this.forEach(itemInArray => {
      // ------------ INSERT CODE HERE! ----------------------------
      // Apply the predicateFunction to each item in the array.
      // If the result is truthy, add the item to the results array.
      // Note: remember you can add items to the array using the array's
      // push() method.
      if (predicateFunction(itemInArray)) {
        results.push(itemInArray);
      }
      // ------------ INSERT CODE HERE! ----------------------------
    });

    return results;
  };

  // JSON.stringify([1,2,3].filter(function(x) { return x > 2})) === "[3]"
}

export function exo8() {
  const newReleases = [
    {
      id: 70111470,
      title: "Die Hard",
      boxart: "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 4.0,
      bookmark: []
    },
    {
      id: 654356453,
      title: "Bad Boys",
      boxart: "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 5.0,
      bookmark: [{ id: 432534, time: 65876586 }]
    },
    {
      id: 65432445,
      title: "The Chamber",
      boxart: "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 4.0,
      bookmark: []
    },
    {
      id: 675465,
      title: "Fracture",
      boxart: "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
      uri: "http://api.netflix.com/catalog/titles/movies/70111470",
      rating: 5.0,
      bookmark: [{ id: 432534, time: 65876586 }]
    }
  ];

  // ------------ INSERT CODE HERE! -----------------------------------
  // Chain the filter and map functions to select the id of all videos
  // with a rating of 5.0.

  //return newReleases; // Complete this expression
  return newReleases
    .filter(video => video.rating === 5.0)
    .map(video => video.id);
  // ------------ INSERT CODE HERE! -----------------------------------
}

export function exo9() {
  const movieLists = [
    {
      name: "New Releases",
      videos: [
        {
          id: 70111470,
          title: "Die Hard",
          boxart: "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 4.0,
          bookmark: []
        },
        {
          id: 654356453,
          title: "Bad Boys",
          boxart: "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 5.0,
          bookmark: [{ id: 432534, time: 65876586 }]
        }
      ]
    },
    {
      name: "Dramas",
      videos: [
        {
          id: 65432445,
          title: "The Chamber",
          boxart: "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 4.0,
          bookmark: []
        },
        {
          id: 675465,
          title: "Fracture",
          boxart: "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 5.0,
          bookmark: [{ id: 432534, time: 65876586 }]
        }
      ]
    }
  ];
  const allVideoIdsInMovieLists = [];

  // ------------   INSERT CODE HERE!  -----------------------------------
  // Use two nested forEach loops to flatten the movieLists into a list of
  // video ids.
  movieLists.forEach(movieList => {
    movieList.videos.forEach(video => {
      allVideoIdsInMovieLists.push(video.id);
    });
  });
  // ------------   INSERT CODE HERE!  -----------------------------------

  return allVideoIdsInMovieLists;
}

export function exo10() {
  Array.prototype.concatAll = function concatAll() {
    let results = [];
    this.forEach(subArray => {
      // ------------ INSERT CODE HERE! ----------------------------
      // Add all the items in each subArray to the results array.
      results = results.concat(subArray);
      // ------------ INSERT CODE HERE! ----------------------------
    });

    return results;
  };

  // JSON.stringify([ [1,2,3], [4,5,6], [7,8,9] ].concatAll()) === "[1,2,3,4,5,6,7,8,9]"
  // [1,2,3].concatAll(); // throws an error because this is a one-dimensional array
}

export function exo11() {
  const movieLists = [
    {
      name: "New Releases",
      videos: [
        {
          id: 70111470,
          title: "Die Hard",
          boxart: "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 4.0,
          bookmark: []
        },
        {
          id: 654356453,
          title: "Bad Boys",
          boxart: "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 5.0,
          bookmark: [{ id: 432534, time: 65876586 }]
        }
      ]
    },
    {
      name: "Dramas",
      videos: [
        {
          id: 65432445,
          title: "The Chamber",
          boxart: "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 4.0,
          bookmark: []
        },
        {
          id: 675465,
          title: "Fracture",
          boxart: "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
          uri: "http://api.netflix.com/catalog/titles/movies/70111470",
          rating: 5.0,
          bookmark: [{ id: 432534, time: 65876586 }]
        }
      ]
    }
  ];

  // ------------   INSERT CODE HERE!  -----------------------------------
  // Use map and concatAll to flatten the movieLists in a list of video ids.
  //  return movieLists; // Complete this expression!
  return movieLists
    .map(movieList => movieList.videos.map(video => video.id))
    .concatAll();
  // ------------   INSERT CODE HERE!  -----------------------------------
}
