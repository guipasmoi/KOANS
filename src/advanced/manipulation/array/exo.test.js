import * as part1 from "./part1";

const buildConsole = () => {
  const console = {
    log: jest.fn(x => x)
  };
  return console;
};

test("exo1", () => {
  const console = buildConsole();
  const result = part1.exo1(console);
  expect(result).toMatchSnapshot();
});

test("exo2", () => {
  const console = buildConsole();
  const result = part1.exo2(console);
  expect([1, 2, 3].map(x => x + 1)).toMatchSnapshot();
});

test("exo3", () => {
  const result = part1.exo3();
  expect(result).toMatchSnapshot();
});

test("exo4", () => {
  part1.exo4();
  expect([1, 2, 3].map(x => x + 1)).toMatchSnapshot();
});

test("exo5", () => {
  const result = part1.exo5();
  expect(result).toMatchSnapshot();
});

test("exo6", () => {
  const result = part1.exo6();
  expect(result).toMatchSnapshot();
});

test("exo7", () => {
  const result = part1.exo7();
  expect([1, 2, 3].filter(x => x > 2)).toMatchSnapshot();
});

test("exo8", () => {
  const result = part1.exo8();
  expect(result).toMatchSnapshot();
});

test("exo9", () => {
  const result = part1.exo9();
  expect(result).toMatchSnapshot();
});

test("exo10", () => {
  part1.exo10();
  expect([[1, 2, 3], [4, 5, 6], [7, 8, 9]].concatAll()).toMatchSnapshot();
});

Array.prototype.concatAll = function concatAll() {
    let results = [];
    this.forEach(subArray => {
        // ------------ INSERT CODE HERE! ----------------------------
        // Add all the items in each subArray to the results array.
        results = results.concat(subArray);
        // ------------ INSERT CODE HERE! ----------------------------
    });
    return results;
};

test("exo11", () => {
    const result = part1.exo11();
    expect(result).toMatchSnapshot();
});
