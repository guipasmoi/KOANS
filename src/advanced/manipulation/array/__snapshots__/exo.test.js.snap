// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exo1 1`] = `
Array [
  Object {
    "id": 70111470,
    "title": "Die Hard",
  },
  Object {
    "id": 654356453,
    "title": "Bad Boys",
  },
  Object {
    "id": 65432445,
    "title": "The Chamber",
  },
  Object {
    "id": 675465,
    "title": "Fracture",
  },
]
`;

exports[`exo2 1`] = `
Array [
  2,
  3,
  4,
]
`;

exports[`exo3 1`] = `
Array [
  Object {
    "id": 70111470,
    "title": "Die Hard",
  },
  Object {
    "id": 654356453,
    "title": "Bad Boys",
  },
  Object {
    "id": 65432445,
    "title": "The Chamber",
  },
  Object {
    "id": 675465,
    "title": "Fracture",
  },
]
`;

exports[`exo4 1`] = `
Array [
  2,
  3,
  4,
]
`;

exports[`exo5 1`] = `
Array [
  Object {
    "id": 70111470,
    "title": "Die Hard",
  },
  Object {
    "id": 654356453,
    "title": "Bad Boys",
  },
  Object {
    "id": 65432445,
    "title": "The Chamber",
  },
  Object {
    "id": 675465,
    "title": "Fracture",
  },
]
`;

exports[`exo6 1`] = `
Array [
  Object {
    "bookmark": Array [
      Object {
        "id": 432534,
        "time": 65876586,
      },
    ],
    "boxart": "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
    "id": 654356453,
    "rating": 5,
    "title": "Bad Boys",
    "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
  },
  Object {
    "bookmark": Array [
      Object {
        "id": 432534,
        "time": 65876586,
      },
    ],
    "boxart": "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
    "id": 675465,
    "rating": 5,
    "title": "Fracture",
    "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
  },
]
`;

exports[`exo7 1`] = `
Array [
  3,
]
`;

exports[`exo8 1`] = `
Array [
  654356453,
  675465,
]
`;

exports[`exo9 1`] = `
Array [
  70111470,
  654356453,
  65432445,
  675465,
]
`;

exports[`exo10 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;

exports[`exo11 1`] = `
Array [
  70111470,
  654356453,
  65432445,
  675465,
]
`;
